client/consensus/slots/src/lib.rs

```diff
@@ -113,6 +113,21 @@ pub trait SimpleSlotWorker<B: BlockT> {
 	/// Returns a `Proposer` to author on top of the given block.
 	fn proposer(&mut self, block: &B::Header) -> Result<Self::Proposer, consensus_common::Error>;
 
+	/// Remaining duration of the slot.
+	fn slot_remaining_duration(&self, slot_info: &SlotInfo) -> Duration {
+		let now = Instant::now();
+		if now < slot_info.ends_at {
+			slot_info.ends_at.duration_since(now)
+		} else {
+			Duration::from_millis(0)
+		}
+	}
+
+	/// Remaining duration for proposing. None means unlimited.
+	fn proposing_remaining_duration(&self, _slot_info: &SlotInfo) -> Option<Duration> {
+		None
```

Comment:
Why not default to `Some(slot_remaining_duration)`? I think it's safer to make sure inner proposal logic is always timed out externally.

Reply:
Okay I changed it to that.

Reply:
Still, I want to note that if we really get a faulty proposer that stuck in an infinite loop, this won't really save us. We currently don't kill offloaded proposing logic. The threadpool has a few spots, if that fills up, no new proposing thread will be created, meaning the offloading future will just always return pending. 

I don't think we need to deal with that extreme situation. If we do, the only (?) solution would probably be to kill those offloaded thread. I currently think that will be a bad idea, as it may poison locks and damage internal state.